## 단원별 심화 연습 문제 (난이도: 중)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn as sns
import glob

# set floating point formatting
pd.options.display.float_format = '{:,.1f}'.format

### Q1

**범위**
- Series, DataFrame 자료구조
- 파일 입출력
- 조회, 정렬, 조건 필터링

`/mnt/elice/dataset/insurance.csv`파일을 읽어서 DataFrame 형식으로 출력합니다. (상위 5개 행을 출력 합니다.)

In [3]:
# 코드를 입력해 주세요
df = pd.read_csv('/mnt/elice/dataset/insurance.csv')
df.head(5)

age     sex  bmi  children smoker     region  charges
0   19  female 27.9         0    yes  southwest 16,884.9
1   18    male 33.8         1     no  southeast  1,725.6
2   28    male 33.0         3     no  southeast  4,449.5
3   33    male 22.7         0     no  northwest 21,984.5
4   32    male 28.9         0     no  northwest  3,866.9

<p><strong>[출력 결과]</strong></p>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>age</th>
      <th>sex</th>
      <th>bmi</th>
      <th>children</th>
      <th>smoker</th>
      <th>region</th>
      <th>charges</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>19</td>
      <td>female</td>
      <td>27.9</td>
      <td>0</td>
      <td>yes</td>
      <td>southwest</td>
      <td>16,884.9</td>
    </tr>
    <tr>
      <th>1</th>
      <td>18</td>
      <td>male</td>
      <td>33.8</td>
      <td>1</td>
      <td>no</td>
      <td>southeast</td>
      <td>1,725.6</td>
    </tr>
    <tr>
      <th>2</th>
      <td>28</td>
      <td>male</td>
      <td>33.0</td>
      <td>3</td>
      <td>no</td>
      <td>southeast</td>
      <td>4,449.5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>33</td>
      <td>male</td>
      <td>22.7</td>
      <td>0</td>
      <td>no</td>
      <td>northwest</td>
      <td>21,984.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>32</td>
      <td>male</td>
      <td>28.9</td>
      <td>0</td>
      <td>no</td>
      <td>northwest</td>
      <td>3,866.9</td>
    </tr>
  </tbody>
</table>
</div>

다음의 조건을 만족하는 DataFrame을 출력하시오.

1. `sex`컬럼이 `male` 인 데이터만 필터합니다.
2. `children` 이 2명 이상인 데이터만 필터합니다.
3. `region`이 `northeast`와 `northwest` 인 데이터만 필터합니다.
4. `bmi`와 `age`를 기준으로 내림차순 정렬하고 3~5번째 행만 선택하여 출력합니다.

In [12]:
# 코드를 입력해 주세요
cond1 = (df['sex'] == 'male') & (df['children'] >= 2)
cond2 = df['region'].isin(['northeast', 'northwest'])
sample = df.loc[cond1 & cond2].sort_values(by=['bmi', 'age'], ascending=[False, False])
sample[3:6]

age   sex  bmi  children smoker     region  charges
569    48  male 40.6         2    yes  northwest 45,702.0
1201   46  male 40.4         2     no  northwest  8,733.2
1318   35  male 39.7         4     no  northeast 19,496.7

<p><strong>[출력 결과]</strong></p>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>age</th>
      <th>sex</th>
      <th>bmi</th>
      <th>children</th>
      <th>smoker</th>
      <th>region</th>
      <th>charges</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>569</th>
      <td>48</td>
      <td>male</td>
      <td>40.6</td>
      <td>2</td>
      <td>yes</td>
      <td>northwest</td>
      <td>45,702.0</td>
    </tr>
    <tr>
      <th>1201</th>
      <td>46</td>
      <td>male</td>
      <td>40.4</td>
      <td>2</td>
      <td>no</td>
      <td>northwest</td>
      <td>8,733.2</td>
    </tr>
    <tr>
      <th>1318</th>
      <td>35</td>
      <td>male</td>
      <td>39.7</td>
      <td>4</td>
      <td>no</td>
      <td>northeast</td>
      <td>19,496.7</td>
    </tr>
  </tbody>
</table>
</div>

다음의 조건을 만족하는 DataFrame을 출력하시오.
1. `sex`컬럼이 `female` 인 데이터만 필터합니다.
2. 흡연자(`smoker` 컬럼이 `yes`)만 필터합니다.
3. 나이는 40대만 필터합니다. (40세 ~ 49세)
4. `charges`를 기준으로 내림차순 정렬하고, 상위 5개 행만 출력 합니다.

In [13]:
# 코드를 입력해 주세요
cond1 = (df['sex'] == 'female') & (df['smoker'] == 'yes')
cond2 = (df['age'] >= 40) & (df['age'] < 50)
sample = df.loc[cond1 & cond2].sort_values(by=['charges'], ascending=False).head(5)
sample

age     sex  bmi  children smoker     region  charges
488    44  female 38.1         0    yes  southeast 48,885.1
674    44  female 43.9         2    yes  southeast 46,201.0
549    43  female 46.2         0    yes  southeast 45,863.2
1323   42  female 40.4         2    yes  southeast 43,896.4
629    44  female 39.0         0    yes  northwest 42,983.5

<p><strong>[출력 결과]</strong></p>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>age</th>
      <th>sex</th>
      <th>bmi</th>
      <th>children</th>
      <th>smoker</th>
      <th>region</th>
      <th>charges</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>488</th>
      <td>44</td>
      <td>female</td>
      <td>38.1</td>
      <td>0</td>
      <td>yes</td>
      <td>southeast</td>
      <td>48,885.1</td>
    </tr>
    <tr>
      <th>674</th>
      <td>44</td>
      <td>female</td>
      <td>43.9</td>
      <td>2</td>
      <td>yes</td>
      <td>southeast</td>
      <td>46,201.0</td>
    </tr>
    <tr>
      <th>549</th>
      <td>43</td>
      <td>female</td>
      <td>46.2</td>
      <td>0</td>
      <td>yes</td>
      <td>southeast</td>
      <td>45,863.2</td>
    </tr>
    <tr>
      <th>1323</th>
      <td>42</td>
      <td>female</td>
      <td>40.4</td>
      <td>2</td>
      <td>yes</td>
      <td>southeast</td>
      <td>43,896.4</td>
    </tr>
    <tr>
      <th>629</th>
      <td>44</td>
      <td>female</td>
      <td>39.0</td>
      <td>0</td>
      <td>yes</td>
      <td>northwest</td>
      <td>42,983.5</td>
    </tr>
  </tbody>
</table>
</div>

다음의 조건을 만족하는 Series를 출력하시오.
1. `bmi`가 30이상 40 미만인 데이터만 필터합니다.
2. `age` 가 30대인 데이터만 필터 합니다.
3. 필터된 데이터의 `region` 컬럼의 분포를 확인합니다.

In [14]:
# 코드를 입력해 주세요
cond1 = (df['bmi'] >= 30) & (df['bmi'] < 40)
cond2 = (df['age'] >= 30) & (df['age'] < 40)
sample = df.loc[cond1 & cond2]['region'].value_counts()
sample

southwest    32
southeast    32
northeast    20
northwest    20
Name: region, dtype: int64

<p><strong>[출력 결과]</strong></p>
<pre>southwest    32
southeast    32
northeast    20
northwest    20
Name: region, dtype: int64</pre>

## 제출

제출을 위해 새로 로드된 `/mnt/elice/dataset/insurance.csv` 데이터에서 아래 조건을 만족하는 결과를 `result_df`에 저장하세요.

1. `sex`컬럼이 `female` 인 데이터만 필터합니다.
2. 흡연자(`smoker` 컬럼이 `yes`)만 필터합니다.
3. 나이는 40대만 필터합니다. (40세 ~ 49세)
4. `charges`를 기준으로 내림차순 정렬하고, 상위 5개 행만 가져옵니다.

In [15]:
df = pd.read_csv('/mnt/elice/dataset/insurance.csv')
cond1 = (df['sex'] == 'female') & (df['smoker'] == 'yes')
cond2 = (df['age'] >= 40) & (df['age'] < 50)
result_df = df.loc[cond1 & cond2].sort_values(by=['charges'], ascending=False).head(5)
result_df

age     sex  bmi  children smoker     region  charges
488    44  female 38.1         0    yes  southeast 48,885.1
674    44  female 43.9         2    yes  southeast 46,201.0
549    43  female 46.2         0    yes  southeast 45,863.2
1323   42  female 40.4         2    yes  southeast 43,896.4
629    44  female 39.0         0    yes  northwest 42,983.5

## 채점 수행

아래 코드는 채점 수행을 위한 코드입니다.

따라서 이를 수정했을시 **채점이 제대로 이루어지지 않습니다.**

**주의**: 채점 코드를 실행하기 전에 반드시 코드 파일을 한번 저장하시길 바랍니다.

In [16]:
import os

result_df.to_json('result.json')

os.system('elice_grade result.json cds_ds_exercise.ipynb')

send files ['result.json', 'cds_ds_exercise.ipynb'] for grade...
waiting result...
waiting result...
waiting result...
done!

Score: 100.000000
Duration: 1.985 seconds
=== Message ===
제출 완료되었습니다.


0